doc/developer-guide/internal-libraries/MemArena.en.rst

```diff
@@ -14,223 +14,102 @@
+done calling :func:`MemArena::thaw` will release the frozen memory blocks.
+
+Other than a freeze / thaw cycle, there is no mechanism to release memory except for the destruction
+of the :class:`MemArena`. This makes :class:`MemArena` not suitable for a container that allocates
+variables sized objects (such as something like a :code:`Std::vector`). A container that allocates
+the same sized objects (such as a :code:`std::map`) can (and should) use a free list.
+
+The :func:`MemArena::make` method is a templated method that provides for constructing object in
+arena memory. A allocation of :code:`size(T)` is made and an instance of :code:`T` constructed in
+it. The method is variadic and will forward all of its arguments to the constructor for :code:`T`.
+The instance must not have :code:`delete` called on it as this will corrupt memory. Because the
+memory for the instance was allocated in the arena that memory will be released when the arena is
+destroyed. The destructor can be called if needed but in general if a destructor is needed it is
+probably not a class that should be constructed in the arena. In particular such objects should not
+own memory that is not also in the arena. Following these rules means the objects need no explicit
+cleanup, destroying the arena suffices. If strings are needed it is best to use
+:code:`std::string_view` and have the view refer to a string allocated in the same arena.
+
+Lists of objects can be handled easily using an instrusive linked list class such as :code:`IntrusiveDList`.
+In this case the objects and the required list infrastructure are all in the arena. That is, while the list
+owns the objects and their memory, all of that is in the arena and therefore cleaned up with the arena.
 
 Reference
 +++++++++
 
 .. class:: MemArena
 
-   .. class:: Block
-      
-      Underlying memory allocated is owned by the :class:`Block`. A linked list. 
-
-      .. member:: size_t size
-      .. member:: size_t allocated
-      .. member:: std::shared_ptr<Block> next
-      .. function:: Block(size_t n)
-      .. function:: char* data()
-
    .. function:: MemArena()
 
       Construct an empty arena.
 
-   .. function:: explicit MemArena(size_t n)
+   .. function:: MemArena(size_t n)
 
-      Construct an arena with :arg:`n` bytes. 
+      Construct an arena with an initial internal block with at least :arg:`n` bytes of free space.
 
    .. function:: MemSpan alloc(size_t n)
 
-      Allocate an :arg:`n` byte chunk of memory in the arena.
+      Allocate memory of size :arg:`n` bytes in the arena.
 
-   .. function:: MemArena& freeze(size_t n = 0)
-
-      Block all further allocation from any existing internal blocks. If :arg:`n` is zero then on the next allocation request a block twice as large as the current generation, otherwise the next internal block will be large enough to hold :arg:`n` bytes.
-
-   .. function:: MemArena& thaw()
+   .. function:: template <typename T> T * make(Args&& ... args)
```

Comment:
```
.. function:: template <typename T, typename... Args> T* make(Args &&... args)
```

Reply:
I tried that, it causes errors in building the docs (`MemArena::make` is not found).

Reply:
I just tried it again, and it works:
```
   .. function:: template < typename T, typename... Args> T * make(Args&& ... args)
```

Here is my output:
```
ts:fedora:28)[~/trafficserver/doc]
linuxkit-025000000001 ∴ make html
make[1]: Entering directory '/Users/derekd/devel/trafficserver/doc'
PAPEROPT_a4="-t latex_a4" PAPEROPT_letter="-t latex_letter" PAPER="letter" PLANTUML_JAR="/Users/derekd/devel/trafficserver/.git/doc-tools/plantuml-1.2018.1/plantuml.jar" ./sbuild sphinx-build -c .  -d docbuild/doctrees -b html . docbuild/html
Running Sphinx v1.7.5
making output directory...
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 488 source files that are out of date
updating environment: 488 added, 0 changed, 0 removed
reading sources... [ 88%] developer-guide/plugins/example-plugins/basic-authorization/implementing-the-handler-and-getting-a-handle-to-the-reading sources... [100%] preface/index.en
/Users/derekd/devel/trafficserver/doc/admin-guide/files/logging.yaml.en.rst:306: WARNING: Malformed table.
Text in column margin in table line 13.

=================== =========== ===============================================
Name                Type        Description
=================== =========== ===============================================
filename            string      The name of the logfile relative to the default
                                logging directory (set with
                                :ts:cv:`proxy.config.log.logfile_dir`).
format              string      a string with a valid named format specification.
header              string      If present, emitted as the first line of each
                                new log file.
rolling_enabled     *see below* Determines the type of log rolling to use (or
                                whether to disable rolling). Overrides
                                :ts:cv:`proxy.config.log.rolling_enabled`.
rolling_interval_sec number     Interval in seconds between log file rolling.
                                Overrides
                                :ts:cv:`proxy.config.log.rolling_interval_sec`.
rolling_offset_hr   number      Specifies an hour (from 0 to 23) at which log
                                rolling is guaranteed to align. Only has an
                                effect if RollingIntervalSec is set to greater
                                than one hour. Overrides
                                :ts:cv:`proxy.config.log.rolling_offset_hr`.
rolling_size_mb     number      Size, in megabytes, at which log files are
                                rolled.
filters             array of    The optional list of filter objects which
                    filters     restrict the individual events logged. The array
                                may only contain one accept filter.
collation_hosts     array of    If present, one or more strings specifying the
                    strings     log collation hosts to which logs should be
                                delivered, each in the form of "<ip>:<port>".
                                :ref:`admin-logging-collation` for more
                                information.
=================== =========== ===============================================
/Users/derekd/devel/trafficserver/doc/admin-guide/layer-4-routing.en.rst:: WARNING: image file not readable: admin-guide/../uml/images/l4-basic-sequence.svg
/Users/derekd/devel/trafficserver/doc/admin-guide/layer-4-routing.en.rst:86: WARNING: image file not readable: admin-guide/../uml/images/l4-example-cdn-layout.svg
/Users/derekd/devel/trafficserver/doc/admin-guide/layer-4-routing.en.rst:: WARNING: image file not readable: admin-guide/../uml/images/l4-sni-routing-seq.svg
/Users/derekd/devel/trafficserver/doc/admin-guide/logging/examples.en.rst:202: WARNING: Malformed table.
Text in column margin in table line 27.

=== ======== ==================================================================
No. Field    Description
=== ======== ==================================================================
1   cqtq     The client request timestamp in Squid format. The time of the
             client request in seconds since January 1, 1970 UTC (with
             millisecond resolution).
2   ttms     The time |TS| spent processing the client request. The number
             of milliseconds between the time the client established the
             connection with |TS| and the time |TS| sent the last byte of
             the response back to the client.
3   chi      The IP address of the client’s host machine.
4   crc/pssc The cache result code; how the cache responded to the request:
             ``HIT``, ``MISS``, and so on. Cache result codes are described
             in :ref:`admin-logging-cache-results`. The
             proxy response status code (HTTP response status code from
             |TS| to client).
5   psql     The length of the |TS| response to the client in bytes,
             including headers and content.
6   cqhm     The client request method: ``GET``, ``POST``, and so on.
7   cauc     The client request canonical URL; blanks and other characters
             that might not be parsed by log analysis tools are replaced by
             escape sequences. The escape sequence is a percentage sign
             followed by the ASCII code number of the replaced character in
             hex.
8   caun     The username of the authenticated client. A hyphen (``-``)
             means that no authentication was required.
9   phr/shn The proxy hierarchy route. The route |TS| used to retrieve the
             object.
10  psct     The proxy response content type. The object content type taken
             from the |TS| response header.
=== ======== ==================================================================
/Users/derekd/devel/trafficserver/doc/developer-guide/core-architecture/rpc.en.rst:69: WARNING: image file not readable: developer-guide/core-architecture/../../uml/images/RPC-sequence-diagram.svg
/Users/derekd/devel/trafficserver/doc/developer-guide/core-architecture/rpc.en.rst:: WARNING: image file not readable: developer-guide/core-architecture/../../uml/images/RPC-states.svg
/Users/derekd/devel/trafficserver/doc/developer-guide/plugins/example-plugins/tls_bridge.en.rst:: WARNING: image file not readable: developer-guide/plugins/example-plugins/../../../uml/images/TLS-Bridge-Plugin.svg
/Users/derekd/devel/trafficserver/doc/developer-guide/plugins/example-plugins/tls_bridge.en.rst:: WARNING: image file not readable: developer-guide/plugins/example-plugins/../../../uml/images/TLS-Bridge-Messages.svg
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [ 88%] developer-guide/plugins/example-plugins/basic-authorization/implementing-the-handler-and-getting-a-handle-to-the-twriting output... [100%] preface/index.en
/Users/derekd/devel/trafficserver/doc/appendices/command-line/traffic_layout.en.rst:34: WARNING: unknown option: --run-root
generating indices... genindex
writing additional pages... search
copying images... [ 22%] admin-guide/monitoring/monitoring/third-party/../../../../static/images/admin/monitor/circonus/new-check-button.pncopying images... [100%] developer-guide/plugins/../../static/images/sdk/txn_sm.jpg
copying static files... done
copying extra files... done
dumping search index in English (code: en) ...
done
dumping object inventory... done
build succeeded, 10 warnings.

The HTML pages are in docbuild/html.

Build finished. The HTML pages are in docbuild/html.
make[1]: Leaving directory '/Users/derekd/devel/trafficserver/doc'
```

This gets rendered for me as:

![image](https://user-images.githubusercontent.com/905298/42226282-33f6d8f4-7ea4-11e8-8945-95dd355f17a7.png)

Maybe this requires a newer sphinx version?


I am thinking it is preferable to document the code signature and return type correctly versus documenting it incorrectly to work around a sphinx bug. But if we can confirm this is not an issue then we don't have to make that choice.

Reply:
@SolidWallOfCode Confirms that with sphinx 1.6.6 there is some parsing issue with this declaration. It seems we have 3 options:

1) require a minimum sphinx version
2) document the code correctly but accept the warning and lack of link
3) document the code incorrectly and have the link work.

These are in my order of preference. What do you think?